analyzer/src/traversal/expressions.rs

```diff
@@ -529,20 +579,162 @@ fn expr_call_builtin_function(
+                name,
+                param_count,
+                ess(param_count),
+                args.kind.len(),
+                if args.kind.len() == 1 { "was" } else { "were" },
+            ),
+            labels,
+            vec![],
+        );
+        // TODO: add `defined here` label (need span for definition)
+    }
+}
+
+pub fn validate_arg_labels(
+    context: Shared<Context>,
+    args: &Node<Vec<Node<fe::CallArg>>>,
+    labels: &[Option<&str>],
+) {
+    for (expected_label, arg) in labels.iter().zip(args.kind.iter()) {
+        let arg_val = &arg.kind.value;
+        match (expected_label, &arg.kind.label) {
+            (Some(expected_label), Some(actual_label)) => {
+                if *expected_label != actual_label.kind {
+                    let notes = if labels
+                        .iter()
+                        .any(|nm| nm == &Some(actual_label.kind.as_str()))
+                    {
+                        vec!["Note: arguments must be provided in order.".into()]
+                    } else {
+                        vec![]
+                    };
+                    context.borrow_mut().fancy_error(
+                        "argument label mismatch",
+                        vec![Label::primary(
+                            actual_label.span,
+                            format!("expected `{}`", expected_label),
+                        )],
+                        notes,
+                    );
+                }
+            }
+            (Some(expected_label), None) => match &arg_val.kind {
+                fe::Expr::Name(var_name) if var_name == *expected_label => {}
+                _ => {
+                    context.borrow_mut().fancy_error(
+                        "missing argument label",
+                        vec![Label::primary(
+                            Span::new(arg_val.span.start, arg_val.span.start),
+                            format!("add `{}=` here", expected_label),
+                        )],
+                        vec![format!(
+                            "Note: this label is optional if the argument is a variable named `{}`.",
+                            expected_label
+                        )],
+                    );
+                }
+            },
+            (None, Some(actual_label)) => {
+                context.borrow_mut().error(
+                    "argument should not be labeled",
+                    actual_label.span,
+                    "remove this label",
+                );
+            }
+            (None, None) => {}
+        }
+    }
+}
+
+pub fn validate_arg_types(
+    scope: Shared<BlockScope>,
+    context: Shared<Context>,
+    name: &str,
+    args: &Node<Vec<Node<fe::CallArg>>>,
+    params: &[(String, FixedSize)],
+) -> Result<(), SemanticError> {
+    for ((label, param_type), arg) in params.iter().zip(args.kind.iter()) {
+        let val_attrs = assignable_expr(Rc::clone(&scope), Rc::clone(&context), &arg.kind.value)?;
```

Comment:
I just noticed that if there are more args than the fn expects, the excess args won't be analyzed. I'm inclined to fix this later.

Reply:
not a big deal